File: 백준/Silver/2578. 빙고/빙고.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];

rl.on("line", (x) => {
  input.push(x);
}).on("close", () => {
  const bingo = [];
  const call = [];
  for (let i = 0; i < 5; i += 1) {
    const line = input[i].split(" ");
    bingo.push(line);
  }

  for (let i = 5; i < input.length; i += 1) {
    const line = input[i].split(" ");
    call.push(...line);
  }

  solution(bingo, call);
});

function solution(bingo, call) {
  let bingoCount = 0;
  const calledBingo = [];
  let result = 0;
  for (let i = 0; bingoCount < 3; i += 1) {
    const idx = findIdx(bingo, call[i]);
    calledBingo.push(idx);
    bingoCount = bingoCheck(calledBingo);
    result = i + 1;
  }
  console.log(result);
}

function findIdx(bingo, num) {
  for (let i = 0; i < bingo.length; i += 1) {
    for (let j = 0; j < bingo[i].length; j += 1) {
      if (bingo[i][j] === num) return [i, j];
    }
  }
}

function bingoCheck(calledBingo) {
  const rowLine = [
    [
      [0, 0],
      [0, 1],
      [0, 2],
      [0, 3],
      [0, 4],
    ],
    [
      [1, 0],
      [1, 1],
      [1, 2],
      [1, 3],
      [1, 4],
    ],
    [
      [2, 0],
      [2, 1],
      [2, 2],
      [2, 3],
      [2, 4],
    ],
    [
      [3, 0],
      [3, 1],
      [3, 2],
      [3, 3],
      [3, 4],
    ],
    [
      [4, 0],
      [4, 1],
      [4, 2],
      [4, 3],
      [4, 4],
    ],
  ];

  const colLine = [
    [
      [0, 0],
      [1, 0],
      [2, 0],
      [3, 0],
      [4, 0],
    ],
    [
      [0, 1],
      [1, 1],
      [2, 1],
      [3, 1],
      [4, 1],
    ],
    [
      [0, 2],
      [1, 2],
      [2, 2],
      [3, 2],
      [4, 2],
    ],
    [
      [0, 3],
      [1, 3],
      [2, 3],
      [3, 3],
      [4, 3],
    ],
    [
      [0, 4],
      [1, 4],
      [2, 4],
      [3, 4],
      [4, 4],
    ],
  ];

  const diagLine = [
    [
      [0, 0],
      [1, 1],
      [2, 2],
      [3, 3],
      [4, 4],
    ],
    [
      [0, 4],
      [1, 3],
      [2, 2],
      [3, 1],
      [4, 0],
    ],
  ];

  let count = 0;

  for (let i = 0; i < colLine.length; i += 1) {
    const include = includesArr(calledBingo, colLine[i]);
    if (include) count += 1;
  }
  for (let i = 0; i < rowLine.length; i += 1) {
    const include = includesArr(calledBingo, rowLine[i]);
    if (include) count += 1;
  }
  for (let i = 0; i < diagLine.length; i += 1) {
    const include = includesArr(calledBingo, diagLine[i]);
    if (include) count += 1;
  }
  return count;
}

function includesArr(mainArr, subArr) {
  const include = subArr.every((sub) =>
    mainArr.some((main) => main.every((val, idx) => val === sub[idx]))
  );
  return include;
}

File: 백준/Silver/2578. 빙고/빙고.py
bingo = []

def bingoCehck(arr):
    count = 0
    #가로 체크
    for i in arr:
        if(i.count(1)==5):
            count+=1
    #세로 체크
    for i in range(5):
        c=0
        for j in range(5):
            if(arr[j][i]==1):
                c+=1
        if(c==5):
            count+=1

    #대각선 체크
    rc = 0
    lc = 0
    for i in range(5):

        if(arr[i][i]==1):
            rc+=1
        if(arr[i][4-i]==1):
            lc+=1
    if(rc==5):
        count+=1
    if(lc==5):
        count+=1

    return count>=3

def findIdx(arr, n):
    for i in range(5):
        for j in range(5):
            if(arr[i][j]==n):
                return [i,j]

for _ in range(5):
    line = list(map(int,input().split()))
    bingo.append(line)

call = []
for _ in range(5):
    line = list(map(int,input().split()))
    for i in line:
        call.append(i)

bingoArr = [[0 for _ in range(5)] for _ in range(5)]
count=0
for i in call:
    count+=1
    x,y = findIdx(bingo, i)
    bingoArr[x][y] = 1
    if(bingoCehck(bingoArr)):
        break

print(count)